src/zkstar.rs

```diff
@@ -0,0 +1,118 @@
+use crate::errors::InternalError;
+use libpaillier::unknown_order::BigNumber;
+use serde::{Deserialize, Serialize};
+
+/// A warpper around the modulus of the group [ZStarN] is defined over
+/// This allows passing the modulus by reference and avoid serializing it
+#[derive(Serialize)]
+struct ZStarNBuilder {
+    N: BigNumber,
+}
+
+impl ZStarNBuilder {
+    pub fn new(n: BigNumber) -> Self {
+        ZStarNBuilder { N: n }
+    }
+    /// The only way to construct an element of [ZStarN] is to validate that [ZStarNUnverified]
+    /// is properly constructed with respect to the current [ZStarNBuilder]
+    pub fn validate(&self, unverified: ZStarNUnverified) -> Result<ZStarN, InternalError> {
+        if unverified.value().is_zero() {
+            return Err(InternalError::IsZero);
+        } else if unverified.value() > self.value() {
+            return Err(InternalError::LargerThanModulus);
+        }
+        let result = unverified.value().extended_gcd(self.value());
+        if result.gcd != BigNumber::one() {
+            return Err(InternalError::NotCoprime);
+        }
+        Ok(ZStarN {
+            value: unverified.value().clone(),
+            builder: self,
+        })
+    }
+    /// Return the modulus of [ZStarNBuilder]
+    fn value(&self) -> &BigNumber {
+        &self.N
+    }
+}
+
+/// An verifier element of [ZStarN]
```

Comment:
suggestion: This is a circular definition ;) I've sprinkled this around in other suggestions, but we _have_ to say "multiplicative group of integers mod `N`" here, at the very least. With `N` defined by the builder.

nit: verifier -> verified.

Reply:
Also, document what the representation is -- I think these are always positive, right? We've been running into some issues with different modules using `[-N/2, N/2]` vs `[0, N)`. I'm not sure which one we actually want here, so just documenting what it does now should be enough to help us debug later if there is a representation problem.

Reply:
Did you mean `(0, N)` ? 0 can't be inclusive.

Reply:
Yes, thanks!